### PR TITLE
[poco] Fix handling of feature NetSSL/NetSSLWin

### LINKED ISSUE #29629: [poco] Fix handling of feature NetSSL/NetSSLWin
**Is your feature request related to a problem? Please describe.**

Recently there have been some modifications of the port (6a57c0e10fa8eafc58f3ff5ce110c82b5a167908) to change the way the modules NetSSL and NetSSLWin are handled. To understand the made changes, it is essential to understand that on Windows you may build either NetSSL or NetSSLWin, but not both. NetSSLWin is not supported on another platform than Windows.

However the way the port works at the moment, it has the following disadvantages:
- It is not possible to use the NetSSL (not NetSSLWin) module on platform Windows. The port always uses NetSSLWin on platform Windows.
- It is not possible to use the Crypto module at all on platform Windows. The port always disables the module Crypto on platform Windows.
- It is not possible to use the Crypto module without using NetSSL on all other platforms. The port enables the module Crypto only and only, if the module NetSSL is enabled (on all platforms other than Windows).

**Proposed solution**
A solution is to declare NetSSLWin instead of NetSSL as a feature (supported on platform Windows only). If that feature is enabled/requested on platform Windows, then NetSSLWin is build _instead_ of NetSSL. Otherwise NetSSL is build on platform Windows. On all other platforms NetSSLWin is never enabled, but NetSSL is. So the proposed solution fulfills the requirement that either NetSSL or NetSSLWin may be enabled on Windows and never enabled NetSSLWin on platforms other than Windows.

Additionally the coupling of the Crypto module to NetSSL/NetSSLWin has to be removed, so that it is build always no matter, if NetSSL/NetSSLWin is build.

### PR DESCRIPTION
**Pull request by i-C-o-d-e-r**
Fixes #29629.

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)
- [ ] SHA512s are updated for each updated download
- [X] The "supports" clause reflects platforms that may be fixed by this new version
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by i-C-o-d-e-r** [comment]
@microsoft-github-policy-service agree

**Review by JonLiu1993 — APPROVED** [review]

**Review by dg0yt — COMMENTED** [review]
I think the change could be implemented like this:
- Add a `netsslwin` feature, `"supports": "windows"`.
- Add a dependency on the new feature to feature `netssl` for `"platform": "windows"`.
- Set options accordingly:

~~~cmake
if("netsslwin" IN_LIST FEATURES)
        list(APPEND FEATURE_OPTIONS
            -DENABLE_CRYPTO=ON
            -DENABLE_NETSSL=OFF
            -DENABLE_NETSSL_WIN=ON
    )
elseif("netssl" IN_LIST FEATURES)
        list(APPEND FEATURE_OPTIONS
            -DENABLE_CRYPTO=ON
            -DENABLE_NETSSL=ON
            -DENABLE_NETSSL_WIN=OFF
    )
else()
        list(APPEND FEATURE_OPTIONS
            -DENABLE_CRYPTO=OFF
            -DENABLE_NETSSL=OFF
            -DENABLE_NETSSL_WIN=OFF
    )
endif()
~~~

File: ports/poco/portfile.cmake [hunk 1]
```diff
@@ -81,6 +78,7 @@ vcpkg_cmake_configure(
         -DENABLE_POCODOC=ON
         -DENABLE_PAGECOMPILER=ON
         -DENABLE_PAGECOMPILER_FILE2PAGE=ON
+        -DENABLE_CRYPTO=ON
```
[the remark this answers is not in the document]
Comment (i-C-o-d-e-r) [reply ports/poco/portfile.cmake:81]: This change was intentional. Issue #29629 describes the problem.

In short: All other modules of the port are also unconditionally enabled. There is no reason why crypto should only be enabled, if netssl or netsslwin is enabled. Crypto is useful without netssl or netsslwin. (I am using it without both)

File: ports/poco/vcpkg.json [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
   "version": "1.12.4",
-  "port-version": 3,
+  "port-version": 4,
   "description": "Modern, powerful open source C++ class libraries for building network and internet-based applications that run on desktop, server, mobile and embedded systems.",
   "homepage": "https://github.com/pocoproject/poco",
   "license": "BSL-1.0",
   "supports": "!uwp",
   "dependencies": [
     "expat",
+    "openssl",
```
[the remark this answers is not in the document]
Comment (i-C-o-d-e-r) [reply ports/poco/vcpkg.json:11]: This is due to the change that "crypto" is enabled unconditionally. In this case openssl is always needed.

File: ports/poco/vcpkg.json [hunk 3]
```diff
@@ -32,14 +33,9 @@
...
-      "description": "NetSSL support for POCO",
-      "dependencies": [
-        {
-          "name": "openssl",
-          "platform": "!windows"
-        }
-      ]
+    "netsslwin": {
+      "description": "NetSSL Windows support for POCO (replaces NetSSL)",
+      "supports": "windows"
```
[the remark this answers is not in the document]
Comment (i-C-o-d-e-r) [reply ports/poco/vcpkg.json:38]: I am not very familiar with requirements in regards to upgrades. Could you try to describe what the problem may be? That would help me to propose a better solution that addresses your concern.

[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/poco/vcpkg.json:38]: Until now, you cannot upgrade an installed `poco[netssl]` if the new version no longer has the feature.

**Comment by i-C-o-d-e-r** [comment]
> I think the change could be implemented like this:
> 
>     * Add a `netsslwin` feature, `"supports": "windows"`.
> 
>     * Add a dependency on the new feature to feature `netssl` for `"platform": "windows"`.
> 
>     * Set options accordingly: ....

#29629 describes what the motivation for this pull request is:

1. Decouple building of `crypto` from `netssl` and/or `netsslwin` (I would like to be able to use `crypto` without `netssl` or `netsslwin`)

2. On Windows you may build either `netssl` or `netsslwin`, but not both

How does you proposal solve this two issues?

**Comment by dg0yt** [comment]
Can you are confirm that code which relies on the NetSSL module will continue to work when the package is switched to the NetSSLWin module?

**Comment by dg0yt** [comment]
> 1. Decouple building of `crypto` from `netssl` and/or `netsslwin` (I would like to be able to use `crypto` without `netssl` or `netsslwin`)

Sound like a case for a `crypto` feature.

**Comment by i-C-o-d-e-r** [comment]
> Sound like a case for a `crypto` feature.

There is no `xml`, `json`, `util` (and so on) feature either. All of those are built unconditionally. While you are right that this should be the ultimate solution, this is not the scope of this pull request. I propose to create a new issue for this proposal.

**Comment by i-C-o-d-e-r** [comment]
> Can you are confirm that code which relies on the NetSSL module will continue to work when the package is switched to the NetSSLWin module?

No I cannot, but that is not the intention. Developers must have the choice on windows, which of both features they would like to use. The current port (without my changes) forces the module netsslwin on windows. There is no way to use netssl on windows. The freedom of choice is what my changes ensure. They do not ensure that you can switch between those two packages later, even though it may be possible (but I don't know for sure).

File: ports/poco/vcpkg.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (i-C-o-d-e-r) [reply ports/poco/vcpkg.json:38]: I understand now, but I have no solution for that.

**Comment by i-C-o-d-e-r** [comment]
I don't see what exactly your proposal is and how it solves the problems I mentioned so far. Could you be more specific please?

**Comment by dg0yt** [comment]
I stopped making proposals and started asking questions.
- Removing a feature is disruptive for explicit downstream usage of this feature. I try to understand if it is necessary at all.
- [Implementing API alternatives must not be done with features.](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#features) I try to understand if the options are strictly alternative or it they can be enabled at the same time.

**Comment by i-C-o-d-e-r** [comment]
Oh, ok now I understand.

Regarding the question, if the options netssl and nettsslwin are strictly alternatives I can tell the following:

It seems that the POCO build system does not prevent that you build both alternatives. However during the installation phase of the header files, header files of one alternative are overwritten with the header files of the other alternative. Depending on which alternative you are using in the source code, that might or might not work.

I encountered this problem in an earlier version of the port and obviously others had the same problem, because 6a57c0e10fa8eafc58f3ff5ce110c82b5a167908 was an explicit fix to the problem.

So to sum it up, my proposed changes are the only way of solving the mentioned problems that I can think of currently.

So as long as no-one has a better approach that solves #29629 and can avoid removing the feature `netssl ` at the same time,...

**Comment by i-C-o-d-e-r** [comment]
@JonLiu1993 Would you please remove the label `requires:author-response`?

Btw: Which author? The autor of the pull request or the author of the port?

**Comment by dg0yt** [comment]
> However during the installation phase of the header files, header files of one alternative are overwritten with the header files of the other alternative (due to same path/filename). Depending on which alternative you are using in the source code, that might or might not work.

These are my findings:
- The ssl options are API alternatives. vcpkg can't offer such alternatives via features. (Alternatives can be implemented via triplet variables or overlay ports.)
- Upstream says NetSSL_Win [is one of the unreleased "use at your own risk" libraries](https://github.com/pocoproject/poco/issues/3706#issuecomment-1195288136). So IMO vcpkg shouldn't select NetSSL_Win.

**Comment by i-C-o-d-e-r** [comment]
> Upstream says NetSSL_Win https://github.com/pocoproject/poco/issues/3706#issuecomment-1195288136. So IMO vcpkg shouldn't select NetSSL_Win.

It would be a feasible option IMO. However this would silently replace NetSSL_Win with NetSSL for all current users of the port.

> Can you are confirm that code which relies on the NetSSL module will continue to work when the package is switched to the NetSSLWin module?

Can you confirm that code which relies on the NetSSLWin module will continue to work when the package is switched to the NetSSL module?

**Comment by dg0yt** [comment]
> Can you confirm that code which relies on the NetSSLWin module will continue to work when the package is switched to the NetSSL module?

No, I can't.  But I don't think an experimental option is a good default for vcpkg.

Maybe upstream made good progress. Does somebody want to ask upstream which alternative vcpkg should choose?

**Comment by i-C-o-d-e-r** [comment]
> No, I can't. But I don't think an experimental option is a good default for vcpkg.

Well, that is exactly what my changes do. My changes ensure that NetSSL is the new default, but allows existing code to explicitly opt-in to NetSSLWin, if it relies on it or needs it. This provides some kind of backwards compatibility for code that relies on the NetSSLWin module, which was what the current port provided as the only alternative at the moment.

> Maybe upstream made good progress. Does somebody want to ask upstream which alternative vcpkg should choose?

IMO https://github.com/pocoproject/poco/issues/3706#issuecomment-1195288136 is clear about NetSSLWin. It should not be used. But it has been the only alternative provided with the port at the moment. So either we remove it completly and risk breaking existing code or we provide a feature/possibility to opt-in to an "use at your own risk library".

But of course, if somebody wants to clearify this with upstream....then please do so.

**Comment by dg0yt** [comment]
> Well, that is exactly what my changes do. My changes ensure that NetSSL is the new default, but allows existing code to explicitly opt-in to NetSSLWin, if it relies on it or needs it. This provides some kind of backwards compatibility for code that relies on the NetSSLWin module, which was what the current port provided as the only alternative at the moment.

So we mostly agree ...  except for whether a feature is the right customization point. 
Do you want me to offer the alternative (triplet variable) as a separate PR? With proper cross-reference, it would offer a fair choice which one to merge.

**Comment by i-C-o-d-e-r** [comment]
> So we mostly agree ... except for whether a feature is the right customization point.

I think there is a misunderstanding here. I totally agree that a feature might not be the right customization point (as you call it). But that is what I was able to do in my limited time now. I have no experience with overlay-ports at all.

> Do you want me to offer the alternative (triplet variable) as a separate PR? With proper cross-reference, it would offer a fair choice which one to merge.

My changes do not only provide an (imperfect) solution to the NetSSL/NetSSLWin problem, but also address the problem of the unwanted coupling of the `crypto` module with NetSSL/NetSSLWin. I would like to see this to be fixed in the port as well.

Therefore would it be a good way ahead, if we merge my PR (as it is) first and then you create another PR that uses overlay-ports/triplets to improve the NetSSL/NetSSLWin issue?

Or you solve the problem with `crypto` in your PR as well, then I have no problem to withdraw my PR.

**Comment by JonLiu1993** [comment]
Duplicate of #29754
